Error 500: unknown token type "\"\"" when working with AppRole

**Describe the bug**

After upgrading to Vault 1.2.1, we can no longer login (or even read the configuration of) one of our AppRole roles. Other roles appear unaffected. We receive the following error:

Error reading auth/approle/role/my-role: Error making API request.

```
URL: GET https://vault.example.com:8200/v1/auth/approle/role/my-role
Code: 500. Errors:

* 1 error occurred:
        * unknown token type "\"\""
```

We've tried to re-write the AppRole with a valid configuration, but get:

```
Error writing data to auth/approle/role/my-role: Error making API request.

URL: PUT https://vault.example.com:8200/v1/auth/approle/role/my-role
Code: 500. Errors:

* 1 error occurred:
        * internal error
```

**To Reproduce**

Steps to reproduce the behavior:
1. Run `vault read auth/approle/role/my-role`
2. See error

**Expected behavior**

Shows configuration for that approle / allows login via approle.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.2.1
* Vault CLI Version (retrieve with `vault version`): 1.2.1
* Server Operating System/Architecture: RHEL 7.7 x86_64

Vault server configuration file(s):

```hcl
storage "consul" {
  address = "127.0.0.1:8500"
  path    = "secrets/"
}

ui = true
cluster_name = "cluster-name"
pid_file = "/var/run/vault.pid"
disable_clustering = true

listener "tcp" {
  address     = "1.2.3.5:8200"
  tls_cert_file = "/etc/pki/tls/certs/vault.crt"
  tls_key_file = "/etc/pki/tls/private/vault.key"
  tls_min_version = "tls12"
  proxy_protocol_behavior = "allow_authorized"
  proxy_protocol_authorized_addrs = "1.2.3.4,2.3.4.5,3.4.5.6"
  tls_disable_client_certs = "true"
}

api_addr = "https://vault-02.example.com:8200"
```

**Additional context**
We upgraded vault from 1.1.3 to 1.2.0 and then 1.2.1. We're not sure if this issue was present as well in 1.2.0 because of the other issues preventing AppRoles from working.

We have a number of other AppRoles, and this one seems to be the only one affected. Nothing really special about it's configuration:

```
{
  "bind_secret_id": true,
  "bound_cidr_list": null,
  "local_secret_ids": false,
  "period": 0,
  "policies": [
    "a-policy",
    "b-policy"
  ],
  "secret_id_bound_cidrs": null,
  "secret_id_num_uses": 0,
  "secret_id_ttl": 0,
  "token_bound_cidrs": null,
  "token_max_ttl": 600,
  "token_num_uses": 0,
  "token_ttl": 600
}
```

Hi @arusso I was able to reproduce this by slightly modifying one of the existing test:

```
=== RUN   TestAppRole_TokenutilUpgrade
--- FAIL: TestAppRole_TokenutilUpgrade (0.00s)
    path_role_test.go:1874: unknown token type "\"\""
```

I'm currently doing a bit more investigation and working on a fix for this.

--- a/builtin/credential/approle/path_role_test.go
+++ b/builtin/credential/approle/path_role_test.go
@@ -1846,16 +1846,25 @@ func createRole(t *testing.T, b *backend, s logical.Storage, roleName, policies 
 func TestAppRole_TokenutilUpgrade(t *testing.T) {
 	tests := []struct {
 		name              string
+		storageValMissing bool
 		storageVal        string
 		expectedTokenType logical.TokenType
 	}{
 		{
+			"token_type_missing",
+			true,
+			"",
+			logical.TokenTypeDefault,
+		},
+		{
 			"token_type_empty",
+			false,
 			"",
 			logical.TokenTypeDefault,
 		},
 		{
 			"token_type_service",
+			false,
 			"service",
 			logical.TokenTypeService,
 		},
@@ -1881,7 +1890,14 @@ func TestAppRole_TokenutilUpgrade(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			entryVal := fmt.Sprintf(`{"policies": ["foo"], "period": 300000000000, "token_bound_cidrs": ["127.0.0.1", "10.10.10.10/24"], "token_type": "%s"}`, tt.storageVal)
+
+			// Construct the storage entry object based on our test case.
+			tokenTypeKV := ""
+			if !tt.storageValMissing {
+				tokenTypeKV = fmt.Sprintf(`, "token_type": "%s"`, tt.storageVal)
+			}
+			entryVal := fmt.Sprintf(`{"policies": ["foo"], "period": 300000000000, "token_bound_cidrs": ["127.0.0.1", "10.10.10.10/24"]%s}`, tokenTypeKV)
+
 			// Hand craft JSON because there is overlap between fields
 			if err := s.Put(ctx, &logical.StorageEntry{
 				Key:   "role/" + tt.name,

--- a/builtin/credential/approle/path_role_test.go
+++ b/builtin/credential/approle/path_role_test.go
@@ -3,6 +3,7 @@ package approle
 import (
 	"context"
 	"encoding/json"
+	"fmt"
 	"reflect"
 	"strings"
 	"testing"
@@ -1843,6 +1844,23 @@ func createRole(t *testing.T, b *backend, s logical.Storage, roleName, policies 
 // TestAppRole_TokenutilUpgrade ensures that when we read values out that are
 // values with upgrade logic we see the correct struct entries populated
 func TestAppRole_TokenutilUpgrade(t *testing.T) {
+	tests := []struct {
+		name              string
+		storageVal        string
+		expectedTokenType logical.TokenType
+	}{
+		{
+			"token_type_empty",
+			"",
+			logical.TokenTypeDefault,
+		},
+		{
+			"token_type_service",
+			"service",
+			logical.TokenTypeService,
+		},
+	}
+
 	s := &logical.InmemStorage{}
 
 	config := logical.TestBackendConfig()
@@ -1861,31 +1879,39 @@ func TestAppRole_TokenutilUpgrade(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Hand craft JSON because there is overlap between fields
-	if err := s.Put(ctx, &logical.StorageEntry{
-		Key:   "role/foo",
-		Value: []byte(`{"policies": ["foo"], "period": 300000000000, "token_bound_cidrs": ["127.0.0.1", "10.10.10.10/24"], "token_type": "service"}`),
-	}); err != nil {
-		t.Fatal(err)
-	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			entryVal := fmt.Sprintf(`{"policies": ["foo"], "period": 300000000000, "token_bound_cidrs": ["127.0.0.1", "10.10.10.10/24"], "token_type": "%s"}`, tt.storageVal)
+			// Hand craft JSON because there is overlap between fields
+			if err := s.Put(ctx, &logical.StorageEntry{
+				Key:   "role/" + tt.name,
+				Value: []byte(entryVal),
+			}); err != nil {
+				t.Fatal(err)
+			}
 
-	fooEntry, err := b.roleEntry(ctx, s, "foo")
-	if err != nil {
-		t.Fatal(err)
-	}
+			resEntry, err := b.roleEntry(ctx, s, tt.name)
+			if err != nil {
+				t.Fatal(err)
+			}
 
-	exp := &roleStorageEntry{
-		SecretIDPrefix: "secret_id/",
-		Policies:       []string{"foo"},
-		Period:         300 * time.Second,
-		TokenParams: tokenutil.TokenParams{
-			TokenPolicies:   []string{"foo"},
-			TokenPeriod:     300 * time.Second,
-			TokenBoundCIDRs: []*sockaddr.SockAddrMarshaler{&sockaddr.SockAddrMarshaler{SockAddr: sockaddr.MustIPAddr("127.0.0.1")}, &sockaddr.SockAddrMarshaler{SockAddr: sockaddr.MustIPAddr("10.10.10.10/24")}},
-			TokenType:       logical.TokenTypeService,
-		},
-	}
-	if diff := deep.Equal(fooEntry, exp); diff != nil {
-		t.Fatal(diff)
+			exp := &roleStorageEntry{
+				SecretIDPrefix: "secret_id/",
+				Policies:       []string{"foo"},
+				Period:         300 * time.Second,
+				TokenParams: tokenutil.TokenParams{
+					TokenPolicies: []string{"foo"},
+					TokenPeriod:   300 * time.Second,
+					TokenBoundCIDRs: []*sockaddr.SockAddrMarshaler{
+						{SockAddr: sockaddr.MustIPAddr("127.0.0.1")},
+						{SockAddr: sockaddr.MustIPAddr("10.10.10.10/24")},
+					},
+					TokenType: tt.expectedTokenType,
+				},
+			}
+			if diff := deep.Equal(resEntry, exp); diff != nil {
+				t.Fatal(diff)
+			}
+		})
 	}
 }

--- a/sdk/logical/token.go
+++ b/sdk/logical/token.go
@@ -38,7 +38,7 @@ func (t *TokenType) UnmarshalJSON(b []byte) error {
 	// Handle upgrade from pre-1.2 where we were serialized as string:
 	s := string(b)
 	switch s {
-	case `"default"`:
+	case `"default"`, `""`:
 		*t = TokenTypeDefault
 	case `"service"`:
 		*t = TokenTypeService

--- a/sdk/logical/token_test.go
+++ b/sdk/logical/token_test.go
@@ -30,4 +30,14 @@ func TestJSONSerialization(t *testing.T) {
 	if tt != utt {
 		t.Fatalf("expected %v, got %v", tt, utt)
 	}
+
+	// Test on an empty value, which should unmarshal into TokenTypeDefault
+	tt = TokenTypeDefault
+	err = json.Unmarshal([]byte(`""`), &utt)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tt != utt {
+		t.Fatalf("expected %v, got %v", tt, utt)
+	}
 }

